File: merge-sort/js/merge-sort.js
function merge(left, right){
  var out = [],
      l = 0,
      r = 0,
      l_len=left.length, 
      r_len=right.length;
  
  while (l < l_len && r < r_len) {
    if (left[l] <= right[r])
      out.push(left[l++]);
    else
      out.push(right[r++]);
  }

  if (l < l_len)
      out = out.concat(left.slice(l));
  else
      out = out.concat(right.slice(r));
  
  return out;
}

function mergeSort(arr) {
  if (arr.length <= 1)
      return arr;

  var mid = Math.floor(arr.length / 2),
      left = arr.slice(0, mid),
      right = arr.slice(mid);

  left = mergeSort(left);
  right = mergeSort(right);
  return merge(left, right);
}

// sanity check
a = [4, 5, 6, 3, 2, 1];
console.log(a + ' --> ' + mergeSort(a));
a = [-3, 0, 3, -4];
console.log(a + ' --> ' + mergeSort(a));
a = ['b', 'a', 'A', 'B'];

File: merge-sort/python/merge_sort.py
def merge(left, right):
    out = []
    l = 0
    r = 0
    l_len = len(left)
    r_len = len(right)
    while l < l_len and r < r_len:
        if left[l] <= right[r]:
            out.append(left[l])
            l += 1
        else:
            out.append(right[r])
            r += 1
    if l < l_len:
        out = out + left[l:]
    else:
        out = out + right[r:]
    return out

def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    mid = int(len(arr)/2)
    left = arr[:mid]
    right = arr[mid:]
    left = merge_sort(left)
    right = merge_sort(right)
    return merge(left, right)

if __name__ == '__main__':
    assert merge_sort([4,5,6,3,2,1]) == [1,2,3,4,5,6]
    assert merge_sort([-3,0,3,-4]) == [-4,-3,0,3]
    assert merge_sort(['z', 'p', 'a', 'd']) == ['a', 'd', 'p', 'z']
    assert merge_sort(['b', 'a', 'A', 'B']) == ['A', 'B', 'a', 'b']
